generator/src/main/java/com/scottlogic/deg/generator/fieldspecs/SetRestrictionsMergeOperation.java

```diff
@@ -1,77 +1,89 @@
+        @Nullable SetRestrictions setRestrictionsB,
+        @NotNull Predicate<Object> valueIsValid) {
 
-            if (!typeRestrictions.isTypeAllowed(IsOfTypeConstraint.Types.DATETIME)) {
-                filterStream = filterStream.filter(x -> !isDateTime(x));
-            }
+        MergeResult<SetRestrictions> mergeResult =
+            setRestrictionsMerger.merge(
+                setRestrictionsA,
+                setRestrictionsB);
 
-            if(merging.getStringRestrictions() != null){
-                filterStream = filterStream.filter(x -> !isString(x) || merging.getStringRestrictions().match(x));
-            }
+        // if the merge result was not successful, it means both mergees had set restrictions and there are no mutually
+        // satisfactory values
+        // (this could happen if we merged, eg, "X in [1, 2]" with either "X in [3, 4]" or "NOT(X in [1, 2])")
+        if (!mergeResult.successful){
+            return SetRestrictions.allowNoValues();
+        }
 
-            if(merging.getNumericRestrictions() != null){
-                filterStream = filterStream.filter(x -> !isNumeric(x) || merging.getNumericRestrictions().match(x));
-            }
+        SetRestrictions mergedSetRestrictions = mergeResult.restrictions;
 
-            if(merging.getDateTimeRestrictions() != null){
-                filterStream = filterStream.filter(x -> !isDateTime(x) || merging.getDateTimeRestrictions().match(x));
-            }
+        // a null merge result means neither of the mergees had any set restrictions, so just return null here
+        if (mergedSetRestrictions == null) {
+            return null;
+        }
 
-            Set<Object> whitelist = filterStream.collect(Collectors.toCollection(HashSet::new));
-            SetRestrictions newSetRestrictions = new SetRestrictions(whitelist,
-                setRestrictions.getBlacklist());
+        // filter down whitelist/blacklist to remove values excluded by other restrictions (eg NumericRestrictions)
+        if (mergedSetRestrictions.getWhitelist() != null) {
+            return SetRestrictions.fromWhitelist(
+                filterSet(
+                    mergedSetRestrictions.getWhitelist(),
+                    valueIsValid));
+        } else {
+            return SetRestrictions.fromBlacklist(
+                filterSet(
+                    mergedSetRestrictions.getBlacklist(),
+                    valueIsValid));
+        }
+    }
 
-            setRestrictions = newSetRestrictions;
+    /** If input is non-null, return set of all values that match the given predicate */
+    private static Set<Object> filterSet(Set<Object> input, Predicate<Object> shouldKeepItem) {
+        if (input == null) {
+            return null;
         }
 
-        return Optional.of(merging.withSetRestrictions(
-            setRestrictions,
-            FieldSpecSource.fromFieldSpecs(left, right)));
+        return input.stream()
+            .filter(shouldKeepItem)
```

Comment:
the context of shouldKeepItem is so lost at this point, i'd rather just call directly onto the method directly.

either by passing the partiallyMerging, or restructure the calls so this filter happens at a higher scope.

Reply:
The context of `shouldKeepItem` is _intentionally_ lost. This function does one thing: Filtering sets according to criteria. Passing in a FieldSpec pollutes that purity.